### PR TITLE
Expand documentation for GUI event processing and exception handling in the testing API

### LINKED ISSUE #1399: Doc: Expand Testing TraitsUI Application Section with how exceptions from GUI are handled
Currently the "Testing TraitsUI Application Section" in the user manual talks little (or none at all) regarding how exceptions from GUI event loops are handled.

It is worth mentioning that where UITester is running the event loop, it also captures the exceptions for them to be reraised to fail a test.

### PR DESCRIPTION
**Pull request by kitchoi**
Closes #1399 

This PR adds a section on GUI event process and exception handling in the documentation for testing TraitsUI applications.

Note to reviewer:
The added file is currently an "orphan". It is intended to be included in the "Discussions" section following the refactoring in #1482.
The content is ready to review. Prior to merging, I will need to make one final change to add a cross reference so that the section appears in the table of content and request one more review for that.

The final change will make the Discussions section in `testing.rst` looks like this (see #1482):
```

Discussions
===========

.. toctree::
   :maxdepth: 1

   testing/discussions/automated_vs_manual_test
   testing/discussions/compatibility_pyface_testing
   testing/discussions/event_loop_and_exceptions
   testing/discussions/target_interaction_location
   testing/discussions/working_of_extensions
```

^^^ The event_loop_and_exceptions is added there. We can also discuss the ordering here.

### CODE REVIEW

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
+GUI event processing and exception handling
+===========================================
+
+The testing API is designed such that GUI event processing is requested
+automatically wherever there may be pending GUI events. While events are being
+processed, the global exception handling is overridden temporarily in order to
+capture any unhandled exceptions. A test error is resulted if any exceptions
+are caught.
```
Comment (aaronayres35) [inline docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:8]: Maybe mention that even if a test error occurs, the event loop is still cleaned up prior to other tests, which prevents test interactions.

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+scope of commands and yields to test instructions every now and then while the
+GUI is still open.
+
+In production environments, some unhandled exceptions are tolerated and
+suppressed by the GUI event loop, but some exceptions could cause the entire
+runtime to abort abnormally. Neither of these conditions are desirable in
+tests. Unhandled exceptions typically indicate bugs and should cause a test to
+error. Abortion of the Python runtime prevents running and collecting
+results from multiple tests in a test suite.
+
```
Comment (aaronayres35) [inline docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:30]: So just to clarify this is the section where we would address the strange behavior when using the tester along side a debugger as mentioned in #1394 (if we do decide to add it)?

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+The testing API is designed such that GUI event processing is requested
+automatically wherever there may be pending GUI events. While events are being
+processed, the global exception handling is overridden temporarily in order to
+capture any unhandled exceptions. A test error is resulted if any exceptions
+are caught.
+
+For example, |UIWrapper.locate|, |UIWrapper.perform| and |UIWrapper.inspect|
+all request pending events to be processed before and/or after interacting
+with GUI states. This automatic behavior is enabled by default, but it can be
+disabled via the |UITester.auto_process_events| flag.
```
Comment (aaronayres35) [inline docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:13]: maybe add a cross reference to the page mentioning use with ModalDialogTester to illustrate a use case for this.  
Seeing this sentence stand alone it might be confusing why I would want to have a gui test that doesn't process any gui events

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+Motivation
+----------
+
+In production environments, the GUI event loop typically blocks as it waits
+and processes new GUI events continuously, until the last window is closed.
+In tests, the GUI event loop needs to be run programmatically for a limited
+scope of commands and yields to test instructions every now and then while the
```
Comment (aaronayres35) [inline docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:21]: nit: should "yields" be "yield"

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+GUI is still open.
+
+In production environments, some unhandled exceptions are tolerated and
+suppressed by the GUI event loop, but some exceptions could cause the entire
+runtime to abort abnormally. Neither of these conditions are desirable in
+tests. Unhandled exceptions typically indicate bugs and should cause a test to
+error. Abortion of the Python runtime prevents running and collecting
+results from multiple tests in a test suite.
```
Comment (aaronayres35) [inline docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:29]: I guess this actually describes the event loop clean up as I mentioned in a comment above.  I don't know why this didnt register on my first pass reading through.  Feel free to disregard the above comment

**Review by aaronayres35 — APPROVED** [review]
LGTM, just a few comments.

As far as ordering of the discussion section I think what you laid out in the PR description seems to make sense to me

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst (hunk 5 above)
Comment (kitchoi) [reply docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:29]: I will mention that at the first paragraph.

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst (hunk 3 above)
Comment (kitchoi) [reply docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:13]: That makes sense. I will add a reference.

File: docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst (hunk 2 above)
Comment (aaronayres35) [reply docs/source/traitsui_user_manual/testing/discussions/event_loop_and_exceptions.rst:30]: SGTM

File: docs/source/traitsui_user_manual/testing.rst [hunk 6]
```diff
@@ -42,11 +42,12 @@ Discussions
 
    testing/discussions/automated_vs_manual_test
    testing/discussions/compatibility_pyface_testing
+   testing/discussions/event_loop_and_exceptions
    testing/discussions/target_interaction_location
    testing/discussions/working_of_extensions
 
-Technical References
-====================
+Technical Reference
```
Comment (kitchoi) [inline docs/source/traitsui_user_manual/testing.rst:49]: Very sneaky drive-by... (guilty)

**Review by aaronayres35 — APPROVED** [review]
Still LGTM! thank you!
